src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,240 @@
+using curlioerr = Interop.libcurl.curlioerr;
+using curliocmd = Interop.libcurl.curliocmd;
+
+namespace System.Net.Http
+{
+    internal partial class CurlHandler
+    {
+        private static readonly char[] s_LineEndings = new char[] { HttpRuleParser.CR, HttpRuleParser.LF };
+        private static readonly string s_LineEndingDelimiter = "\r\n";
+        private ulong CurlReceiveHeadersCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context)
+        {
+            return ExecuteCallback(buffer, size, nitems, context,
+                (ptr, totalSize, state) =>
+                {
+                    state.CancellationToken.ThrowIfCancellationRequested();
+
+#if DEBUG
+                    LogCurlBuffer(ptr, totalSize);
+#endif
+
+                    // The callback is invoked once per header and multi-line
+                    // headers get merged into a single line
+                    string responseHeader = Marshal.PtrToStringAnsi(ptr).Trim(s_LineEndings).Trim();
+                    HttpResponseMessage response = state.ResponseMessage;
+
+                    // If this is the final header, we are done
+                    if (string.IsNullOrWhiteSpace(responseHeader))
+                    {
+                        state.ResponseMessage.SignalHeadersReceived();
+                        return totalSize;
+                    }
+
+                    // TODO: Understand scenarios where multiple sets of headers are received
+                    if (!TryParseStatusLine(response, responseHeader))
+                    {
+                        int colonIndex = responseHeader.IndexOf(':');
+
+                        // Skip malformed header lines that are missing the colon character.
+                        if (colonIndex > 0)
+                        {
+                            string headerName = responseHeader.Substring(0, colonIndex);
+                            string headerValue = responseHeader.Substring(colonIndex + 1);
+
+                            if (!response.Headers.TryAddWithoutValidation(headerName, headerValue))
+                            {
+                                // TODO: Skip compression related headers
+                                response.Content.Headers.TryAddWithoutValidation(headerName, headerValue);
+                            }
+                        }
+                    }
+                    return totalSize;
+                });
+        }
+
+        private ulong CurlReceiveBodyCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context)
+        {
+            return ExecuteCallback(buffer, size, nitems, context,
+                (ptr, totalSize, state) =>
+                {
+#if DEBUG
+                    LogCurlBuffer(ptr, totalSize);
+#endif
+
+                    // TODO: Handle compression and chunked-mode
+
+                    var byteBuffer = new byte[totalSize];
```

Comment:
It's unfortunate if for every response we need to allocate a byte buffer to store the entire response.  I'd prefer to see this implemented with a fixed-sized buffer, e.g. 8192 bytes or something like that, that's used to copy from ptr to contentStream in a loop, and then that standard-sized buffer can be pooled so that it can be long-lived and reused across many responses.  @davidsh, @CIPop, and @pgavlin may have other ideas.